### PR TITLE
Move cast level and spell variant info to origin flag

### LINKED ISSUE #2122: Add heightened spell level to chat-card and attack roll message flags.
In GitLab by @william.ring.91 on Mar 1, 2022, 10:26

For use in parsing messages for data, the heightened spell level will allow for automation of damage rolls without relying on looking at the flavor text. Use case: in xdy-workbench

### PR DESCRIPTION
**Pull request by oWave**
There are currently 3 different versions of `flags.pf2e.origin` for storing/retrieving an item:
- ChatMessage contains the full item data, correctly applying spell overlays (variant/heighten) when retrieving the item
- Check/Damage only stores the item uuid, making it impossible to fetch the full item with that message alone
- `SpellPF2e.createTemplate` has its own completely different format

This PR is my attempt at unifying the creation of (and retrieval from) `flags.pf2e.origin`. This would fix #2122 (to check another box in #4882).

### CODE REVIEW

**Comment by stwlam** [comment]
Where is #2122 not satisfied?

**Comment by CarlosFdez** [comment]
I'll have to check, but spell templates using their own data is actually a module compat thing for the prettifying modules. I think it'd be wise to extend it for the templates to add additional stuff like traits (or break everything and tell people to change things, that's always an option). 

What's the compatibility with spells casted from the final scroll? The reason items are sometimes full embedded is because consuming a scroll gets rid of it, thus making the item lookup fail. Same for templates for spells made from a scroll that has been consumed, is it still possible to get the traits and name of the spell for that template?

**Comment by CarlosFdez** [comment]
Good timing honestly, since the best time to break all this is 5.0 which is almost around the corner. I'll take a look with a more fine toothed comb as time permits.

**Comment by CarlosFdez** [comment]
Sorry for the late response. I needed to think for a while on this one, and I realize now why I was uncertain.

Doing it this way assumes that spells will always know how to recreate themselves *and* their spellcasting entry even into the forseeable future. In the future however, a stave will be able to load a spell via a uuid and cast it directly. Shoving casting data as well means that we'll have to ensure the spell knows how it was cast, perhaps via clone and flag injection shenaniganry.

Origin is for retrieving the item but casting is for retrieving the spellcasting entry. The one area where this distinction gets kinda fuzzy is casting#level (aka castLevel), but is absolutely required for the spell to recreate itself (whereas the rest of casting just figures out the spellcasting entry to use).

Also we don't need those item static helpers, keeping those to chat message might be best.

This may be more longer term than you're willing to commit to pf2e as a volunteer, but for now I'd recommend moving spellVariant and casting.level to origin.variant and origin.level). Then later, having damage create the casting flags directly if the item.isOfType("spell").

**Comment by oWave** [comment]
I have simplified the changes. I'll leave touching spell casting to somebody else (or future me)

> a stave will be able to load a spell via a uuid and cast it directly  

That would also solve the issue with casting from consumables. I'd like to try that next (in another PR)

Should the creation of/from origin flags (at least for spell) be done in helper functions to reduce duplication? I've realized that an item getter on MeasuredTemplate would be helpful, but not a fan of having that code around twice.

**Comment by CarlosFdez** [comment]
I think the item non-static method that you subclass to get origin flags was fine, its the other 2 static ones (that were only called by chat message) I had issues with. I'll take a look soon.

I'll likely want a change for spell variant but let me think about it. castLevel will need to change for remaster soon anyways so we can keep that one (otherwise I'd say just level is fine, we could determine what level means based on context).

**Comment by CarlosFdez** [comment]
@stwlam I'm thinking of merging this but renaming `spellVariantOverlayIds?: string[];` to `variant?: { overlays: string[] }` after I do. Thoughts?

File: src/module/chat-message/data.ts [hunk 1]
```diff
@@ -11,18 +11,24 @@ interface ChatMessageSourcePF2e extends foundry.documents.ChatMessageSource {
     flags: ChatMessageFlagsPF2e;
 }
 
+export interface ItemOriginFlag {
+    type: ItemType;
+    uuid: string;
+    castLevel?: number;
+    spellVariantOverlayIds?: string[];
```
Comment (CarlosFdez) [inline src/module/chat-message/data.ts:18]:
```suggestion
    variant?: { overlays: string[] };
```

File: src/module/chat-message/document.ts [hunk 2]
```diff
@@ -120,9 +120,9 @@ class ChatMessagePF2e extends ChatMessage {
         }
 
         if (item?.isOfType("spell")) {
-            const spellVariant = this.flags.pf2e.spellVariant;
-            const castLevel = this.flags.pf2e.casting?.level ?? item.level;
-            const modifiedSpell = item.loadVariant({ overlayIds: spellVariant?.overlayIds, castLevel });
+            const overlayIds = this.flags.pf2e.origin?.spellVariantOverlayIds;
```
Comment (CarlosFdez) [inline src/module/chat-message/document.ts:123]:
```suggestion
            const overlayIds = this.flags.pf2e.origin?.variant?.overlays;
```

File: src/module/item/spell/document.ts [hunk 3]
```diff
@@ -865,6 +857,13 @@ class SpellPF2e<TParent extends ActorPF2e | null = ActorPF2e | null> extends Ite
         );
     }
 
+    override getOriginData(): ItemOriginFlag {
+        const flag = super.getOriginData();
+        flag.castLevel = this.level;
+        if (this.isVariant) flag.spellVariantOverlayIds = [...this.appliedOverlays!.values()];
```
Comment (CarlosFdez) [inline src/module/item/spell/document.ts:863]:
```suggestion
        if (this.isVariant && this.appliedOverlays) {
            flag.variant = { overlays: [...this.appliedOverlays.values()] };
        }

```

File: src/module/scene/measured-template-document.ts [hunk 4]
```diff
@@ -1,9 +1,28 @@
...
+> extends MeasuredTemplateDocument<TParent> {
+    get item(): ItemPF2e<ActorPF2e> | null {
+        const origin = this.flags.pf2e?.origin as Record<string, unknown>;
+        const uuid = origin?.uuid;
+        if (!uuid) return null;
+        const item = fromUuidSync(uuid as string);
+        if (!(item instanceof ItemPF2e)) return null;
+
+        if (item?.isOfType("spell")) {
+            const overlayIds = origin?.spellVariantOverlayIds as string[];
```
Comment (CarlosFdez) [inline src/module/scene/measured-template-document.ts:17]:
```suggestion
            const overlayIds = origin?.variant?.overlays;
```

File: src/module/scene/measured-template-document.ts [hunk 5]
```diff
@@ -1,9 +1,28 @@
...
 import { ScenePF2e } from "./document.ts";
+import { ItemPF2e } from "@item";
+import { ActorPF2e } from "@actor";
 
 export class MeasuredTemplateDocumentPF2e<
     TParent extends ScenePF2e | null = ScenePF2e | null
-> extends MeasuredTemplateDocument<TParent> {}
+> extends MeasuredTemplateDocument<TParent> {
+    get item(): ItemPF2e<ActorPF2e> | null {
+        const origin = this.flags.pf2e?.origin as Record<string, unknown>;
```
Comment (CarlosFdez) [inline src/module/scene/measured-template-document.ts:10]:
```suggestion
        const origin = this.flags.pf2e?.origin;
```

**Pull request merged**